### PR TITLE
Fix converting caption shortcode with link

### LINKED ISSUE #12310: Conversion from Classic to Blocks: caption shortcode fails to convert correctly
Tested with the captioned image shortcode from the themes unit test data:

```
[caption id="attachment_754" align="alignnone" width="604"]<a href="http://build.wordpress-develop.test/wp-content/uploads/2011/07/100_5478.jpg"><img class="wp-image-754 size-large" src="http://build.wordpress-develop.test/wp-content/uploads/2011/07/100_5478.jpg?w=604" alt="Bell on Wharf" width="604" height="453" /></a> Bell on wharf in San Francisco[/caption]
```

- open a post with the content above in Gutenberg (adjust the attachment ID if needed)
- content is in a Classic block, as expected:

screenshot 2018-11-26 at 11 33 47

- in the Classic block "More Options" menu, click "Convert to Blocks"
- the captioned image gets converted to an Image block **with two image**s: the second image is within the caption:

screenshot 2018-11-26 at 11 50 37

Code after conversion:
```
<!-- wp:image {"id":754,"align":"none"} -->
<figure class="wp-block-image alignnone"><a href="http://build.wordpress-develop.test/wp-content/uploads/2011/07/100_5478.jpg"><img src="http://build.wordpress-develop.test/wp-content/uploads/2011/07/100_5478.jpg?w=604" alt="Bell on Wharf" class="wp-image-754"/></a><figcaption><a href="http://build.wordpress-develop.test/wp-content/uploads/2011/07/100_5478.jpg"><img class="wp-image-754 size-large" src="http://build.wordpress-develop.test/wp-content/uploads/2011/07/100_5478.jpg?w=604" alt="Bell on Wharf" width="604" height="453" /></a> Bell on wharf in San Francisco</figcaption></figure>
<!-- /wp:image -->
```

### PR DESCRIPTION
**Pull request by ellatrix**
## Description
Avoids a duplicate image when converting from a classic block.

Fixes #12310.

### Testing Instructions
- Create a classic post
- Insert an image.
- Add a caption, AND link to the media file
- Publish the classic post
- Edit the post in the block editor
- Convert the classic block to blocks
screen shot 2018-12-09 at 11 49 27 am
- In master we see:
screen shot 2018-12-09 at 11 48 14 am
In the branch we see:
screen shot 2018-12-09 at 11 51 14 am

## How has this been tested?
See #12310. E2e tests have been added.

## Types of changes
Bug fix
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @iseulde this generally worked correctly in my tests. I think there is only one attribute missing from this conversion. We are not setting the linkDestination attribute.
So for example, when I convert the following data:
```
[caption id="attachment_6" align="alignnone" width="300"]<a href="http://www.example.com"><img class="wp-image-6 size-medium" src="http://enchanting-caterpillar.w5.poopy.life/wp-content/uploads/2018/11/Nov-15-2018-12-08-48-300x240.gif" alt="" width="300" height="240" /></a> aaaaa[/caption]
```

The link to field goes from "Custom URL" in the classic editor to "None" but if I change "Link to" to custom URL to correct custom link appears.

File: packages/block-library/src/image/index.js [hunk 1]
```diff
@@ -186,7 +186,7 @@ export const settings = {
 					caption: {
 						shortcode: ( attributes, { shortcode } ) => {
 							const { content } = shortcode;
-							return content.replace( /\s*<img[^>]*>\s/, '' );
+							return content.replace( /\s*(?:<a[^>]*>\s*)?<img[^>]*>(?:\s*<\/a>)?\s*/, '' );
```
Comment (mcsf) [inline packages/block-library/src/image/index.js]: Could we handle this with more confidence by natively parsing HTML, instead of using RegExp?

Comment (gwwar) [reply packages/block-library/src/image/index.js]: What's our target browsers? We can use the browser's dom parser to parse
```
if ( typeof DOMParser !== 'undefined' && DOMParser.prototype.parseFromString ) {
	const parser = new DOMParser();
	const parsed = parser.parseFromString( html, 'text/html' );
}
```

Comment (gwwar) [reply packages/block-library/src/image/index.js]: Ah https://developer.mozilla.org/en-US/docs/Web/API/DOMParser has an example of a try catch fallback to regex, if it's not available.

**Comment by alexsanford** [comment]
I think we're missing a few more attributes as well:

- `width`
- `height`
- `linkDestination` (mentioned above)
- `linkTarget`

`width` and `height` may technically be out of scope for this PR, but we should probably get the link attributes working at least.

**Comment by gwwar** [comment]
@alexsanford Did you mean custom width and height? I verified the missing link to the media file, but I choose a large size which appeared to transfer over on block conversion.

File: packages/block-library/src/image/index.js (hunk 1 above)
Comment (dmsnell) [reply packages/block-library/src/image/index.js]: noting that all of the supported browsers in `@wordpress/browserslist-config` provide the `DOMParser()`

Comment (gwwar) [reply packages/block-library/src/image/index.js]: Spinning up an example alternative

**Comment by alexsanford** [comment]
@gwwar Yeah, the custom width and height attributes on the `img` tag. I'm using the "Custom Size" option when editing the Media in the classic editor. Then, in Gutenberg, when I convert to blocks and then look at the width and height (either in the block sidebar or by looking at the HTML), the custom size does not seem to carry over.

**Comment by gwwar** [comment]
Note that captions can contain arbitrary html

**Comment by iseulde** [comment]
What is `linkDestination`?

**Comment by gwwar** [comment]
@iseulde Looks like you beat me :) taking this for another spin.

**Comment by gwwar** [comment]
> What is linkDestination?

Preserving these options:
screen shot 2018-12-09 at 2 04 58 pm

**Comment by iseulde** [comment]
How do we source it from the example here:

https://github.com/WordPress/gutenberg/pull/12315#pullrequestreview-178401010

File: packages/block-library/src/image/index.js (hunk 1 above)
Comment (iseulde) [reply packages/block-library/src/image/index.js]: Oh sorry, didn't see the last comments.

**Comment by gwwar** [comment]
> How do we source it from the example here:

If there's a link, the first child is an a tag wrapping an image. The href in that tag should be the one we want. Eg:

screen shot 2018-12-09 at 2 15 25 pm

**Comment by gwwar** [comment]
Manually tests well with plain captions, and captions with links and styling 👍 . The missing link attribute is maybe a separate issue. I'd be okay with a follow up PR if this isn't the right spot to modify that.

**Comment by iseulde** [comment]
> I'd be okay with a follow up PR if this isn't the right spot to modify that.

We'd need to get the media and attachment link from the ID, then compare it. I agree that this can be a separate PR.

**Review by gwwar — APPROVED** [review]
Thanks @iseulde, let's go with this then! Do you mind adding another issue for the follow up bug, if we don't have one already?

**Review by jorgefilipecosta — APPROVED** [review]
Thank you for iterating on this, In my tests, these changes did not cause any regressions, and make the current behavior better 👍
There are still some improvements that can be done, but they are a good fit for another follow-up PR.

**Comment by jorgefilipecosta** [comment]
To note that I'm not totally sure if this PR totally fixes https://github.com/WordPress/gutenberg/issues/12310.

Even in this branch if I convert the following shortcode:
```
[caption id="attachment_20816" align="alignnone" width="300"]<a href="http://yah.local/wp-content/uploads/2018/12/Oct-29-2018-19-09-47-1.gif"><img class="wp-image-20816 size-medium" src="http://yah.local/wp-content/uploads/2018/12/Oct-29-2018-19-09-47-1-300x223.gif" alt="" width="300" height="223" /></a> A[/caption]
````
I get an image block with another inline image in the caption (it shows two images).
I removed the auto fixing of the issue so we can continue to look the issue further now being able to take advantage of these improvements.

**Comment by jorgefilipecosta** [comment]
I removed comment as my test environment for the case I referred was wrong I think this totally addresses https://github.com/WordPress/gutenberg/issues/12310.

**Pull request merged**